File: Geometry/Edge/eta.cpp
#include "eta.h"

#include <stdio.h>

typedef map<TriPosition, Eta*, TriPositionCompare> EtaIndex;
static EtaIndex* Index = NULL;

Eta::Eta( Edge& e ){}

void Eta::recalculate(){
   // DEFAULT
   value = 1;     
}

void Eta::remove() {
     deleteDependents();
     Index->erase(pos);
     delete this;
}

Eta::~Eta(){}

Eta* Eta::At( Edge& e ){
  TriPosition T( 1, e.getSerialNumber() );
  if( Index == NULL ) Index = new EtaIndex();
  EtaIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Eta* val = new Eta( e );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Eta::CleanUp(){
  if( Index == NULL ) return;
  EtaIndex::iterator iter;
  EtaIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

void Eta::print(FILE* out) {
  map<int, Edge>::iterator eit;
  
  fprintf(out, "Eta [ e ]\n=============\n");
  for(eit = Triangulation::edgeTable.begin(); eit != Triangulation::edgeTable.end(); eit++) {
     fprintf(out, "Eta [%3d]\t= % 2.8f\n", eit->first, Eta::valueAt(eit->second));
  }       
}

File: Java/src/geoquant/Eta.java
package geoquant;

import java.util.HashMap;

import triangulation.Edge;


public class Eta extends Geoquant {
  private static HashMap<TriPosition, Eta> Index = new HashMap<TriPosition, Eta>();
  
  private Eta(Edge e) {
    super(e); // ALWAYS have to call this first
    
  }
  
  public static Eta at(Edge e) {
    TriPosition T = new TriPosition( e.getSerialNumber() );
    Eta q = Index.get(T);
    if(q == null) {
      q = new Eta(e);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Edge e) {
    return at(e).getValue();
  }
  
  protected void recalculate() {
    // This is empty for an eta
  }

 
  public void remove() {
    deleteDependents();
    Index.remove(pos);
  }

}